src/Test/Plutip/Internal/LocalCluster.hs

```diff
@@ -256,3 +220,36 @@ launchChainIndex (RunningNode sp _block0 (_gp, _vData)) dir = do
           }
   void . async $ void $ ChainIndex.runMain config chainIndexConfig
   return $ cicPort chainIndexConfig
+
+data ClusterStatus (a :: Type)
+  = ClusterStarting
+  | ClusterStarted a
+  | ClusterClosing
+  | ClusterClosed
+
+{- | Starting a cluster with a setup action
+ We're heavily depending on cardano-wallet local cluster tooling, however they don't allow the
+ start and stop actions to be two separate processes, which is needed for tasty integration.
+ Instead of rewriting and maintaining these, I introduced a semaphore mechanism to keep the
+ cluster alive until the ClusterClosing action is called.
+-}
+startCluster :: forall (a :: Type). ReaderT ClusterEnv IO a -> IO (TVar (ClusterStatus a), a)
+startCluster onClusterStart = do
+  status <- newTVarIO ClusterStarting
+  void $
+    forkFinally
+      ( withPlutusInterface' $ \clusterEnv -> do
+          res <- runReaderT onClusterStart clusterEnv
+          atomically $ writeTVar status (ClusterStarted res)
+          atomically $ readTVar status >>= \case ClusterClosing -> pure (); _ -> retrySTM
```

Comment:
Maybe
`\case ClusterClosing -> pure (); _ -> retrySTM`
can be shortened with `checkSTM`

Reply:
I actually couldn't use `checkSTM` because in order to compare `ClusterStatus a`, I'd need to add the Eq constraint to `a` (Contract result), which I wanted to avoid.